lib/msf/core/exploit/remote/ldap.rb

```diff
@@ -135,5 +135,107 @@ def discover_base_dn(ldap)
+      if ldap.nil?
+        raise LDAPConnectionNotSpecified.new('No LDAP connection was specfied for bind_to_server_and_get_dn to utilize!')
+      end
+
+      bind_result = ldap.as_json['result']['ldap_result']
+
+      # Codes taken from https://ldap.com/ldap-result-code-reference-core-ldapv3-result-codes
+      case bind_result['resultCode']
+      when 0
+        print_good('Successfully bound to the LDAP server!')
+      when 1
+        raise OperationsError.new("An operational error occurred, perhaps due to lack of authorization. The error was: #{bind_result['errorMessage']}")
+      when 7
+        raise AuthMethodNotSupported.new('Target does not support the simple authentication mechanism!')
+      when 8
+        raise StrongerAuthRequired.new("Server requires a stronger form of authentication than we can provide! The error was: #{bind_result['errorMessage']}")
+      when 14
+        raise SASLBindInProgress.new("Server requires additional information to complete the bind. Error was: #{bind_result['errorMessage']}")
+      when 48
+        raise InappropriateAuthentication.new("Target doesn't support the requested authentication type we sent. Try binding to the same user without a password, or providing credentials if you were doing anonymous authentication.")
+      when 49
+        raise InvalidCredentials.new('Invalid credentials provided!')
+      else
+        raise UnknownError.new("Unknown error occurred whilst binding: #{bind_result['errorMessage']}")
+      end
+
+      if base_dn
+        print_good("Using already discovered base DN: #{base_dn}")
+      elsif base
+        print_good("Using user specified base DN: #{base}")
+        base_dn = base
+      else
+        print_status('Discovering base DN automatically')
+
+        unless (base_dn = discover_base_dn(ldap))
+          print_warning("Couldn't discover base DN!")
+          return nil
+        end
+      end
+
+      if base_prefix.blank?
+        full_base_dn = base_dn.to_s
+      else
+        full_base_dn = "#{base_prefix},#{base_dn}"
+      end
+
+      return full_base_dn, base_dn
+    end
+
+    def check_query_result_code(query_result, filter)
+      if query_result.class != Hash
+        print_error('Parameter to check_query_result_code was not a Hash!')
+        return nil
+      end
+
+      case query_result['resultCode']
+      when 0
+        vprint_good('Successfully queried LDAP server!')
+      when 1
+        print_error("Could not perform query #{filter}. Its likely the query requires authentication!")
+        fail_with(Msf::Exploit::Remote::Failure::NoAccess, query_result['errorMessage'])
+      else
+        if query_result['errorMessage'].blank?
+          fail_with(Msf::Exploit::Remote::Failure::UnexpectedReply, "Query #{filter} failed but no error message was returned!")
+        else
+          fail_with(Msf::Exploit::Remote::Failure::UnexpectedReply, "Query #{filter} failed with error: #{query_result['errorMessage']}")
```

Comment:
mixin methods should not `fail_with` please return something that the consuming module can decide how to handle.

Reply:
Fixed with https://github.com/rapid7/metasploit-framework/pull/17122/commits/9950e7ee6dd5d904e268408bb476460fcc47f8fb where I moved this `fail_with` logic into the modules vs the libraries.